dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jclouds/JCloudsStore.java

```diff
@@ -0,0 +1,221 @@
+
+    fileStoreConfig = new FileStoreConfig(provider, location, container);
+
+    String identity = configurationProvider.getProperty(ConfigurationKey.FILESTORE_IDENTITY);
+    String secret = configurationProvider.getProperty(ConfigurationKey.FILESTORE_SECRET);
+    blobStoreContext =
+        ContextBuilder.newBuilder(provider)
+            .credentials(identity, secret)
+            .overrides(configureOverrides(provider, endpoint))
+            .build(BlobStoreContext.class);
+  }
+
+  private String validateProvider(String provider) {
+    if (!SUPPORTED_PROVIDERS.contains(provider)) {
+      throw new IllegalArgumentException(
+          "Configuration contains unsupported file store provider '"
+              + provider
+              + "'. Falling back to file system provider instead.");
+    }
+
+    if (JCLOUDS_PROVIDER_KEY_FILESYSTEM.equals(provider)
+        && !locationManager.externalDirectorySet()) {
+      throw new IllegalArgumentException(
+          "File system file store provider could not be configured; external directory is not set. ");
+    }
+
+    return provider;
+  }
+
+  private Properties configureOverrides(String provider, String endpoint) {
+    if (JCLOUDS_PROVIDER_KEY_FILESYSTEM.equals(provider)
+        && locationManager.externalDirectorySet()) {
+      Properties overrides = new Properties();
+      overrides.setProperty(
+          FilesystemConstants.PROPERTY_BASEDIR, locationManager.getExternalDirectoryPath());
+      return overrides;
+    }
+
+    if (JCLOUDS_PROVIDER_KEY_AWS_S3.equals(provider)) {
+      Properties overrides = new Properties();
+      overrides.setProperty(S3Constants.PROPERTY_S3_VIRTUAL_HOST_BUCKETS, "false");
+      return overrides;
+    }
+
+    if (JCLOUDS_PROVIDER_KEY_S3.equals(provider)) {
+      Properties overrides = new Properties();
+      overrides.setProperty(S3Constants.PROPERTY_S3_VIRTUAL_HOST_BUCKETS, "false");
+
+      if (StringUtils.isNotEmpty(endpoint)) {
+        overrides.setProperty(PROPERTY_ENDPOINT, endpoint);
+      }
+      return overrides;
+    }
+
+    return new Properties();
+  }
+
+  @PostConstruct
+  public void init() {
+    Location location = createLocation(fileStoreConfig.provider, fileStoreConfig.location);
+    blobStoreContext.getBlobStore().createContainerInLocation(location, fileStoreConfig.container);
+
+    log.info(
+        "File store configured with provider: '{}', container: '{}' and location: '{}'.",
+        fileStoreConfig.provider,
+        fileStoreConfig.container,
+        fileStoreConfig.location);
+  }
+
+  @PreDestroy
+  public void cleanUp() {
+    blobStoreContext.close();
+  }
+
+  private static Location createLocation(String provider, String location) {
+    if (location == null) {
+      // some BlobStores allow specifying a location, such as US-EAST, where containers will exist.
+      // null will choose a default location.
+      return null;
```

Comment:
This will satisfy the scenario outlined in the comment. I wonder could it cause issues in other scenarios? If null is not intended but passed in and a null location is retrieved somewhere else later in the code. Might be an edge case but just flagging it.